### ISSUE TITLE
Standalone && and || trigger infinite loop

### ISSUE THREAD

**Issue by Lydxn** [berry-lang/berry#371]
Either of these two lines will trigger an infinite loop, where you would expect them to do nothing:

```berry
0 && 1
1 || 0
```

This behavior seems to magically resolve when the expression is assigned to a variable, e.g.

```berry
_ = 0 && 1
```

**Comment by s-hadinger** [comment]
I cannot reproduce:

```
> 0 && 1
false
> 1 || 0
true
> def f() 0 && 1 end
syntax_error: stdin:1: strict: expression without side effect detected
> def f() return 0 && 1 end
> f()
false
```

**Comment by skiars** [comment]
These lines need to be parsed together. Try `0 && 1 1 || 0` in REPL.

**Comment by s-hadinger** [comment]
Ah yes indeed. What's interesting is when running in strict mode, it gets the following error (which is expected):
```
> 0 && 1 1 || 0
syntax_error: stdin:1: strict: expression without side effect detected
```

Now looking at the produced code:
```
> def f() 0 && 1 1 || 0 end
> import debug
> debug.codedump(f)
source 'stdin', function 'f':
; line 1
  0000  LDCONST	R0	K0
  0001  JMPF	R0	#0001
  0002  LDCONST	R0	K1
  0003  JMPT	R0	#0003
  0004  RET	0
```

There is definitely a problem at line 3.

**Comment by s-hadinger** [comment]
I'm closing since it's definitely a pathological example that never showed up in any code. Btw using strict mode raises an error to prevent such problem from arising.

**Comment by JayXon** [comment]
here's a normal looking code that reproduces this issue
```berry
for i:1..100
 i%3 || print(i)
end
```

**Comment by s-hadinger** [comment]
Thanks. I'm not sure I would hire somebody writing such code ;)
But I see the bug. Be aware it's a low priority bug.

**Comment by s-hadinger** [comment]
This is fixed in #495

**Comment by JayXon** [comment]
Thanks for fixing this, I can confirm that works now, however I noticed that a slightly tweaked version is giving me syntax error
```berry
for i:1..100
 (i+1)%3 || print(i)
end
```
```
syntax_error: string:2: 'i' undeclared (first use in this function)
```

**Comment by s-hadinger** [comment]
Wow, good catch. Actually it is a completely different bug, since `for i:1..100 (i) end` gives the same error. It looks to be linked that it is enclosed in parenthesis.
I will look at it.

**Comment by s-hadinger** [comment]
The problem is actually a syntax issue. Berry ignores spaces during parsing which means that is parses like this:
```berry
# source code
for i:1..100 (i) end

# how it is interpreted
for i:(1..100)(i) end
```

The problem is that the open parenthesis is considered as a function call over the iterator `(1..100)`, and since it tries to call with `i` parameter, the error is caused because `i` is not yet in the scope.

Overall it's a global syntax issue, because Berry accepts `f (x)` as equivalent to `f(x)` for function calls.

I quick and ugly fix is to use semicolo `;` to separate statements.

```berry
for i:1..100;
 (i+1)%3 || print(i)
end
```

**Comment by s-hadinger** [comment]
@skiars I don't like the fact that `f (x)` is identical to `f(x)` and considers as a function call.

Would you be ok to introduce a new token type in addition to `OptLBK` to indicate that the open parenthesis is touching an identifier? Like `OptLBKCall`

This is a change to the syntax but I believe it's a good change

**Comment by s-hadinger** [comment]
This last bug is fixed in https://github.com/berry-lang/berry/pull/497

**Comment by skiars** [comment]
I agree with the concern. The fact that `f (x)` is treated exactly the same as `f(x)` means the call operator ignores whitespace, which feels like an early design mistake and makes the syntax more ambiguous than it needs to be. In JavaScript, newlines have semantic meaning and can help disambiguate, but Berry treats all whitespace the same, so we don’t even have that tool available.

In this particular case, `for i:1..5 (i > 0) || print(i) end` actually results in a parse error (albeit a confusing one), rather than silently doing the wrong thing. And adding an explicit `;` like:
```berry
for i:0..10;
  (i%3) || print(i)
end

for i : (0..10); (i%3) || print(i) end
```

doesn’t look *that* bad to me, especially since `(i%3) || print(i)` itself is already a bit strange. I’m not sure yet how many similar corner cases exist, so I’m hesitant to say we must forbid spaces before `(` for function calls. It probably deserves a more careful, global look at how Berry wants to treat whitespace and statement boundaries.

That said, it might still be feasible to adjust this part of the language. All the ugly ambiguities revolve around one thing — whether a `(` starts a parenthesized expression or a function call. If we want Berry code to freely use spaces and newlines without surprising interactions, then imposing a restriction here (e.g., requiring no space before `(` for calls) could be a justified trade-off.

**Comment by s-hadinger** [comment]
I'm advocating towards the language preventing you from making mistakes as much as possible.

This could happen when using ternary operator, and would be hard to find:
```berry
class myclass
  # whatever
  def dosomething()
    self.method1            # the dev forgot to call the method
    (self.a == nil) ? self.a = 1 : self.dosomethin()
  end
end
```

Hence the code would be interpreted as:
```berry
class myclass
  # whatever
  def dosomething()
    self.method1(self.a == nil) ? self.a = 1 : self.dosomethin()   # which is a completely different semantic
  end
end
```

In the case above, even `strict` mode would not raise an error because the forgotten parameters are mistakenly taken from the next line.

Anyways, @skiars I have proposed a fix for this: https://github.com/berry-lang/berry/pull/497

I tested it with the entire codebase of Tasmota Berry code, there was no impact - which makes it a decent large scale test.
